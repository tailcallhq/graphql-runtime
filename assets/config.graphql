schema @server(port: "8000", baseURL: "http://jsonplaceholder.typicode.com") {
  query: Query
}

type Query {
  posts: [Post] @http(path: "/posts")
}

type Post {
  id: Int!
  userId: Int!
  title: String!
  body: String!
}