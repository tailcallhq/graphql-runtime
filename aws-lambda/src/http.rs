use std::sync::Arc;

use anyhow::Result;
use hyper::body::Bytes;
use reqwest::Client;
use tailcall::http::Response;
use tailcall::HttpIO;

#[derive(Clone)]
pub struct LambdaHttp {
    client: Client,
}

impl Default for LambdaHttp {
    fn default() -> Self {
        Self { client: Client::new() }
    }
}

impl LambdaHttp {
    pub fn init() -> Self {
        Default::default()
    }
}

#[async_trait::async_trait]
impl HttpIO for LambdaHttp {
    async fn execute(&self, request: reqwest::Request) -> Result<Response<Bytes>> {
        let req_str = format!("{} {}", request.method(), request.url());
        let response = self.client.execute(request).await?.error_for_status()?;
        let res = Response::from_reqwest(response).await?;
        tracing::info!("{} {}", req_str, res.status.as_u16());
        Ok(res)
    }
}

pub fn to_request(
    req: lambda_http::Request,
) -> Result<hyper::Request<hyper::Body>, lambda_http::http::Error> {
    hyper::Request::builder()
        .method(req.method())
        .uri(req.uri())
        .body(hyper::Body::from(req.body().to_vec()))
}

pub fn init_http() -> Arc<LambdaHttp> {
    Arc::new(LambdaHttp::init())
}
