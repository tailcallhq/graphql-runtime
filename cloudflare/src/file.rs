use std::rc::Rc;

use anyhow::anyhow;
use tailcall::FileIO;
use worker::Env;

use crate::r2_address::R2Address;
use crate::to_anyhow;

#[derive(Clone)]
pub struct CloudflareFileIO {
  env: Rc<Env>,
}

impl CloudflareFileIO {
  pub fn init(env: Rc<Env>) -> Self {
    CloudflareFileIO { env }
  }
}

impl CloudflareFileIO {
  async fn bucket(&self, r2: &R2Address) -> anyhow::Result<worker::Bucket> {
    Ok(self.env.bucket(&r2.bucket).map_err(|e| anyhow!(e.to_string()))?)
  }

  async fn get(&self, r2: &R2Address) -> anyhow::Result<String> {
    log::debug!("Reading from bucket:{} path:{}", r2.bucket, r2.path);
    let bucket = self.bucket(&r2).await.map_err(to_anyhow)?;

    let maybe_object = bucket.get(&r2.path).execute().await.map_err(to_anyhow)?;
    let object = maybe_object.ok_or(anyhow!("File {} was not found in bucket: {}", r2.path, r2.bucket))?;

    let body = match object.body() {
      Some(body) => body.text().await.map_err(to_anyhow),
      None => Ok("".to_string()),
    };
    body
  }

  async fn put(&self, r2: &R2Address, value: Vec<u8>) -> anyhow::Result<()> {
    let bucket = self.bucket(&r2).await.map_err(to_anyhow)?;
    bucket.put(&r2.path, value).execute().await.map_err(to_anyhow)?;
    Ok(())
  }
}

impl FileIO for CloudflareFileIO {
  async fn write<'a>(&'a self, file_path: &'a str, content: &'a [u8]) -> anyhow::Result<()> {
    let r2 = R2Address::from_string(file_path.to_string())?;
    self.put(&r2, content.to_vec()).await.map_err(to_anyhow)?;
    Ok(())
  }

  async fn read<'a>(&'a self, file_path: &'a str) -> anyhow::Result<String> {
    let r2 = R2Address::from_string(file_path.to_string())?;
    let content = self.get(&r2).await.map_err(to_anyhow)?;
    Ok(content)
  }
}
