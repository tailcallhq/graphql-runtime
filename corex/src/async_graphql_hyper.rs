use std::any::Any;

use anyhow::Result;
use async_graphql::{BatchResponse, Executor};
use hyper::header::{HeaderValue, CACHE_CONTROL, CONTENT_TYPE};
use hyper::{Body, Response, StatusCode};
use once_cell::sync::Lazy;
use serde::{Deserialize, Serialize};

#[async_trait::async_trait]
pub trait GraphQLRequestLike {
  fn data<D: Any + Clone + Send + Sync>(self, data: D) -> Self;
  async fn execute<E>(self, executor: &E) -> GraphQLResponse
  where
    E: Executor;
}

#[derive(Debug, Deserialize)]
pub struct GraphQLBatchRequest(pub async_graphql::BatchRequest);
impl GraphQLBatchRequest {}

#[async_trait::async_trait]
impl GraphQLRequestLike for GraphQLBatchRequest {
  fn data<D: Any + Clone + Send + Sync>(mut self, data: D) -> Self {
    for request in self.0.iter_mut() {
      request.data.insert(data.clone());
    }
    self
  }
  /// Shortcut method to execute the request on the executor.
  async fn execute<E>(self, executor: &E) -> GraphQLResponse
  where
    E: Executor,
  {
    GraphQLResponse(executor.execute_batch(self.0).await)
  }
}

#[derive(Debug, Deserialize)]
pub struct GraphQLRequest(pub async_graphql::Request);

impl GraphQLRequest {}

#[async_trait::async_trait]
impl GraphQLRequestLike for GraphQLRequest {
  #[must_use]
  fn data<D: Any + Send + Sync>(mut self, data: D) -> Self {
    self.0.data.insert(data);
    self
  }
  /// Shortcut method to execute the request on the schema.
  async fn execute<E>(self, executor: &E) -> GraphQLResponse
  where
    E: Executor,
  {
    GraphQLResponse(executor.execute(self.0).await.into())
  }
}

#[derive(Debug, Serialize)]
pub struct GraphQLResponse(pub async_graphql::BatchResponse);
impl From<async_graphql::BatchResponse> for GraphQLResponse {
  fn from(batch: async_graphql::BatchResponse) -> Self {
    Self(batch)
  }
}
impl From<async_graphql::Response> for GraphQLResponse {
  fn from(res: async_graphql::Response) -> Self {
    Self(res.into())
  }
}

impl From<GraphQLQuery> for GraphQLRequest {
  fn from(query: GraphQLQuery) -> Self {
    let mut request = async_graphql::Request::new(query.query);

    if let Some(operation_name) = query.operation_name {
      request = request.operation_name(operation_name);
    }

    if let Some(variables) = query.variables {
      let value = serde_json::from_str(&variables).unwrap_or_default();
      let variables = async_graphql::Variables::from_json(value);
      request = request.variables(variables);
    }

    GraphQLRequest(request)
  }
}

#[derive(Debug)]
pub struct GraphQLQuery {
  query: String,
  operation_name: Option<String>,
  variables: Option<String>,
}

impl GraphQLQuery {
  /// Shortcut method to execute the request on the schema.
  pub async fn execute<E>(self, executor: &E) -> GraphQLResponse
  where
    E: Executor,
  {
    let request: GraphQLRequest = self.into();
    request.execute(executor).await
  }
}

static APPLICATION_JSON: Lazy<HeaderValue> = Lazy::new(|| HeaderValue::from_static("application/json"));

impl GraphQLResponse {
  pub fn to_response(self) -> Result<Response<hyper::Body>> {
    let mut response = Response::builder()
      .status(StatusCode::OK)
      .header(CONTENT_TYPE, APPLICATION_JSON.as_ref())
      .body(Body::from(serde_json::to_string(&self.0)?))?;

    if self.0.is_ok() {
      if let Some(cache_control) = self.0.cache_control().value() {
        response
          .headers_mut()
          .insert(CACHE_CONTROL, HeaderValue::from_str(cache_control.as_str())?);
      }
    }

    Ok(response)
  }

  /// Sets the `cache_control` for a given `GraphQLResponse`.
  ///
  /// The function modifies the `GraphQLResponse` to set the `cache_control` `max_age`
  /// to the specified `min_cache` value and `public` flag to `cache_public`
  ///
  /// # Arguments
  ///
  /// * `res` - The GraphQL response whose `cache_control` is to be set.
  /// * `min_cache` - The `max_age` value to be set for `cache_control`.
  /// * `cache_public` - The negation of `public` flag to be set for `cache_control`.
  ///
  /// # Returns
  ///
  /// * A modified `GraphQLResponse` with updated `cache_control` `max_age` and `public` flag.
  pub fn set_cache_control(mut self, min_cache: i32, cache_public: bool) -> GraphQLResponse {
    match self.0 {
      BatchResponse::Single(ref mut res) => {
        res.cache_control.max_age = min_cache;
        res.cache_control.public = cache_public;
      }
      BatchResponse::Batch(ref mut list) => {
        for res in list {
          res.cache_control.max_age = min_cache;
          res.cache_control.public = cache_public;
        }
      }
    };
    self
  }
}
