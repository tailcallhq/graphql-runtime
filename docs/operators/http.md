---
title: "@http"
---

This **@http** operator serves as an indication of a field or node that is underpinned by a REST API. For Example:

```graphql showLineNumbers
type Query {
  users: [User] @http(path: "/users")
}
```

In this example, the `@http` operator is added to the `users` field of the `Query` type. This means that the `users` field is underpinned by a REST API. The [path](#path) argument is used to specify the path of the REST API. In this case, the path is `/users`. This means that the GraphQL server will make a GET request to `https://jsonplaceholder.typicode.com/users` when the `users` field is queried.

## baseURL

This refers to the base URL of the API. If not specified, the default base URL is the one specified in the [@upstream](#upstream) operator.

```graphql showLineNumbers
type Query {
  users: [User] @http(path: "/users", baseURL: "https://jsonplaceholder.typicode.com")
}
```

## path

This refers to the API endpoint you're going to call. For instance https://jsonplaceholder.typicode.com/users`.

```graphql showLineNumbers
type Query {
  users: [User] @http(path: "/users")
}
```

If your API endpoint contains dynamic segments, you can use Mustache templates to substitute variables. For example, to fetch a specific user, the path can be written as `/users/{{args.id}}`.

```graphql showLineNumbers
type Query {
  user(id: ID!): User @http(path: "/users/{{args.id}}")
}
```

## method

This refers to the HTTP method of the API call. Commonly used methods include GET, POST, PUT, DELETE, etc. If not specified, the default method is GET. For example:

```graphql showLineNumbers
type Mutation {
  createUser(input: UserInput!): User @http(method: "POST", path: "/users")
}
```

## query

This represents the query parameters of your API call. You can pass it as a static object or use Mustache template for dynamic parameters. These parameters will be added to the URL. For example:

```graphql showLineNumbers
type Query {
  userPosts(id: ID!): [Post] @http(path: "/posts", query: [{key: "userId", value: "{{args.id}}"}])
}
```

## body

The body of the API call. It's used for methods like POST or PUT that send data to the server. You can pass it as a static object or use a Mustache template to substitute variables from the GraphQL variables. For example:

```graphql showLineNumbers
type Mutation {
  createUser(input: UserInput!): User @http(method: "POST", path: "/users", body: "{{args.input}}")
}
```

In the example above, the `createUser` mutation sends a POST request to `/users`, with the input object converted to JSON and included in the request body.

## headers

The `headers` parameter allows you to customize the headers of the HTTP request made by the `@http` operator. It is used by specifying a key-value map of header names and their values.

For instance:

```graphql showLineNumbers
type Mutation {
  createUser(input: UserInput!): User @http(path: "/users", headers: [{key: "X-Server", value: "Tailcall"}])
}
```

In this example, a request to `/users` will include an additional HTTP header `X-Server` with the value `Tailcall`.

You can make use of mustache templates to provide dynamic values for headers, derived from the arguments or [context] provided in the request. For example:

[context]: /docs/guides/context

```graphql showLineNumbers
type Mutation {
  users(name: String): User
    @http(path: "/users", headers: [{key: "X-Server", value: "Tailcall"}, {key: "User-Name", value: "{{args.name}}"}])
}
```

In this scenario, the `User-Name` header's value will dynamically adjust according to the `name` argument passed in the request.

## groupBy

The `groupBy` parameter groups multiple data requests into a single call. For more details please refer out [n + 1 guide].

[n + 1 guide]: /docs/guides/n+1#solving-using-batching

```graphql showLineNumbers
type Post {
  id: Int!
  name: String!
  user: User @http(path: "/users", query: [{key: "id", value: "{{value.userId}}"}], groupBy: ["id"])
}
```

- `query: {key: "id", value: "{{value.userId}}"}]`: Here, TailCall CLI is instructed to generate a URL where the user id aligns with the `userId` from the parent `Post`. For a batch of posts, the CLI compiles a single URL, such as `/users?id=1&id=2&id=3...id=10`, consolidating multiple requests into one.
