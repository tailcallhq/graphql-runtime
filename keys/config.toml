#TODO: add script to generate TLS cert and key; for development purposes run the following command:
# openssl req -x509 -newkey rsa:4096 -keyout keys/key.pem -out keys/cert.pem -days 365 -nodes -subj "/CN=localhost"
[grpc_server]
host = "127.0.0.1"
port = 50051
tls_cert_file = "cert.pem"
tls_key_file = "key.pem"
