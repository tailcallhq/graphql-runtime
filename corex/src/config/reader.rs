use futures_util::future::join_all;
use url::Url;

use crate::config::{Config, Source};
use crate::{FileIO, HttpIO};

/// Reads the configuration from a file or from an HTTP URL and resolves all linked assets.
pub struct ConfigReader<File, Http> {
  file: File,
  http: Http,
}

struct FileRead {
  content: String,
  path: String,
}

impl<File: FileIO, Http: HttpIO> ConfigReader<File, Http> {
  pub fn init(file: File, http: Http) -> Self {
    Self { file, http }
  }

  /// Reads a file from the filesystem or from an HTTP URL
  async fn read_file<T: ToString>(&self, file: T) -> anyhow::Result<FileRead> {
    // Is an HTTP URL
    let content = if let Ok(url) = Url::parse(&file.to_string()) {
      let response = self
        .http
        .execute(reqwest::Request::new(reqwest::Method::GET, url))
        .await?;

      String::from_utf8(response.body.to_vec())?
    } else {
      // Is a file path
      self.file.read(&file.to_string()).await?
    };

    Ok(FileRead { content, path: file.to_string() })
  }

  /// Reads all the files in parallel
  async fn read_files<T: ToString>(&self, files: &[T]) -> anyhow::Result<Vec<FileRead>> {
    let files = files.iter().map(|x| self.read_file(x.to_string()));
    let content = join_all(files).await.into_iter().collect::<anyhow::Result<Vec<_>>>()?;
    Ok(content)
  }

  pub async fn read<T: ToString>(&self, files: &[T]) -> anyhow::Result<Config> {
    let files = self.read_files(files).await?;
    let mut config = Config::default();
    for file in files.iter() {
      let source = Source::detect(&file.path)?;
      let schema = &file.content;
      let new_config = Config::from_source(source, schema)?;
      config = config.merge_right(&new_config);
    }

    Ok(config)
  }
}
