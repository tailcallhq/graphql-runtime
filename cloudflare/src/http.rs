use anyhow::Result;
use reqwest::Client;
use reqwest_middleware::{ClientBuilder, ClientWithMiddleware};
use tailcall::http::Response;
use tailcall::io::HttpIO;

#[derive(Clone)]
pub struct HttpCloudflare {
  client: ClientWithMiddleware,
}

impl Default for HttpCloudflare {
  fn default() -> Self {
    Self { client: ClientBuilder::new(Client::new()).build() }
  }
}

impl HttpCloudflare {
  pub fn init() -> Self {
    let client = ClientBuilder::new(Client::new());
    Self { client: client.build() }
  }
}

#[async_trait::async_trait]
impl HttpIO for HttpCloudflare {
  // HttpClientOptions are ignored in Cloudflare
  // This is because there is little control over the underlying HTTP client
  async fn execute_raw(&self, request: reqwest::Request) -> Result<Response<Vec<u8>>> {
    let client = self.client.clone();
    async_std::task::spawn_local(async move {
      let response = client.execute(request).await?;
      Response::from_reqwest(response).await
    })
    .await
  }
}
