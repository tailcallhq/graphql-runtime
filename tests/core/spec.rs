extern crate core;

use std::collections::BTreeMap;
use std::path::Path;
use std::sync::Arc;
use std::{fs, panic};

use anyhow::Context;
use colored::Colorize;
use futures_util::future::join_all;
use http::Request;
use hyper::Body;
use serde::{Deserialize, Serialize};
use tailcall::core::app_context::AppContext;
use tailcall::core::async_graphql_hyper::{GraphQLBatchRequest, GraphQLRequest};
use tailcall::core::blueprint::{Blueprint, BlueprintError};
use tailcall::core::config::reader::ConfigReader;
use tailcall::core::config::transformer::Required;
use tailcall::core::config::{Config, ConfigModule, Source};
use tailcall::core::http::handle_request;
use tailcall::core::print_schema::print_schema;
use tailcall::core::variance::Invariant;
use tailcall_prettier::Parser;
use tailcall_valid::{Cause, Valid, ValidationError, Validator};

use super::file::File;
use super::http::Http;
use super::model::*;
use super::runtime::ExecutionSpec;
use crate::core::runtime;

#[derive(Debug, Default, Deserialize, Serialize, PartialEq)]
struct SDLError {
    message: String,
    trace: Vec<String>,
    description: Option<String>,
}

impl<'a> From<Cause<&'a str>> for SDLError {
    fn from(value: Cause<&'a str>) -> Self {
        SDLError {
            message: value.message.to_string(),
            trace: value.trace.iter().map(|e| e.to_string()).collect(),
            description: None,
        }
    }
}

impl From<Cause<String>> for SDLError {
    fn from(value: Cause<String>) -> Self {
        SDLError {
            message: value.message.to_string(),
            trace: value.trace.iter().map(|e| e.to_string()).collect(),
            description: value.description,
        }
    }
}

async fn is_sdl_error(spec: &ExecutionSpec, config_module: Valid<ConfigModule, String>) -> bool {
    if spec.sdl_error {
        // errors: errors are expected, make sure they match
        let blueprint = config_module.and_then(|cfg| match Blueprint::try_from(&cfg) {
            Ok(blueprint) => Valid::succeed(blueprint),
            Err(e) => Valid::from_validation_err(BlueprintError::to_validation_string(e)),
        });

        match blueprint.to_result() {
            Ok(_) => {
                tracing::error!("\terror FAIL");
                panic!(
                    "Spec {} {:?} with \"sdl error\" directive did not have a validation error.",
                    spec.name, spec.path
                );
            }
            Err(error) => {
                let errors: Vec<SDLError> =
                    error.as_vec().iter().map(|e| e.to_owned().into()).collect();

                let snapshot_name = format!("{}_error", spec.safe_name);

                insta::assert_json_snapshot!(snapshot_name, errors);
            }
        };

        return true;
    }
    false
}

async fn check_identity(spec: &ExecutionSpec) {
    // TODO: we should probably figure out a way to do this for every test
    // but GraphQL identity checking is very hard, since a lot depends on the code
    // style the re-serializing check gives us some of the advantages of the
    // identity check too, but we are missing out on some by having it only
    // enabled for either new tests that request it or old graphql_spec
    // tests that were explicitly written with it in mind
    if spec.check_identity {
        for (source, content) in spec.server.iter() {
            if matches!(source, Source::GraphQL) {
                let config = Config::from_source(source.to_owned(), content).unwrap();
                let actual = config.to_sdl();

                // \r is added automatically in windows, it's safe to replace it with \n
                let content = content.replace("\r\n", "\n");

                let path_str = spec.path.display().to_string();
                let context = format!("path: {}", path_str);

                let actual = tailcall_prettier::format(actual, &tailcall_prettier::Parser::Gql)
                    .await
                    .map_err(|e| e.with_context(context.clone()))
                    .unwrap();

                let expected = tailcall_prettier::format(content, &tailcall_prettier::Parser::Gql)
                    .await
                    .map_err(|e| e.with_context(context.clone()))
                    .unwrap();

                pretty_assertions::assert_eq!(
                    actual,
                    expected,
                    "Identity check failed for {:#?}",
                    spec.path,
                );
            } else {
                panic!(
                    "Spec {:#?} has \"check identity\" enabled, but its config isn't in GraphQL.",
                    spec.path
                );
            }
        }
    }
}

async fn run_query_tests_on_spec(
    spec: ExecutionSpec,
    server: Vec<ConfigModule>,
    mock_http_client: Arc<Http>,
) {
    if let Some(tests) = spec.test.as_ref() {
        let app_ctx = spec
            .app_context(
                server.first().unwrap(),
                spec.env.clone().unwrap_or_default(),
                mock_http_client.clone(),
            )
            .await;

        // test: Run test specs

        for (i, test) in tests.iter().enumerate() {
            let response = run_test(app_ctx.clone(), test)
                .await
                .context(spec.path.to_str().unwrap().to_string())
                .unwrap();

            let mut headers: BTreeMap<String, String> = BTreeMap::new();

            for (key, value) in response.headers() {
                headers.insert(key.to_string(), value.to_str().unwrap().to_string());
            }

            let response: APIResponse = APIResponse {
                status: response.status().clone().as_u16(),
                headers,
                body: Some(APIBody::Value(
                    serde_json::from_slice(
                        &hyper::body::to_bytes(response.into_body()).await.unwrap(),
                    )
                    .unwrap_or_default(),
                )),
            };

            let snapshot_name = format!("{}_{}", spec.safe_name, i);

            insta::assert_json_snapshot!(snapshot_name, response);
        }

        mock_http_client.test_hits(&spec.path);
    }
}

async fn test_spec(spec: ExecutionSpec) {
    // insta settings
    let mut insta_settings = insta::Settings::clone_current();
    insta_settings.set_prepend_module_to_snapshot(false);
    let _guard = insta::Settings::bind_to_scope(&insta_settings);

    let mock_http_client = Arc::new(Http::new(&spec));

    let mut runtime = runtime::create_runtime(mock_http_client.clone(), spec.env.clone(), None);
    runtime.file = Arc::new(File::new(spec.clone()));
    let reader = ConfigReader::init(runtime);

    // Resolve all configs
    let config_modules = join_all(spec.server.iter().map(|(source, content)| async {
        let config = Config::from_source(source.to_owned(), content)?;

        reader.resolve(config, spec.path.parent()).await
    }))
    .await;

    let config_module = Valid::from_iter(config_modules.iter(), |config_module| {
        Valid::from(config_module.as_ref().map_err(|e| {
            match e.downcast_ref::<ValidationError<String>>() {
                Some(err) => err.clone(),
                None => ValidationError::new(e.to_string()),
            }
        }))
    })
    .and_then(|cfgs| {
        cfgs.into_iter()
            .fold(Valid::succeed(ConfigModule::default()), |acc, c| {
                acc.and_then(|acc| acc.unify(c.clone()))
            })
    })
    // Apply required transformers to the configuration
    .and_then(|cfg| cfg.transform(Required));

    // check sdl error if any
    if is_sdl_error(&spec, config_module.clone()).await {
        return;
    }

    let merged = config_module.to_result().unwrap().to_sdl();

    let formatter = tailcall_prettier::format(merged, &Parser::Gql)
        .await
        .unwrap();

    let snapshot_name = format!("{}_merged", spec.safe_name);

    insta::assert_snapshot!(snapshot_name, formatter);

    let config_modules = config_modules
        .into_iter()
        .collect::<Result<Vec<_>, _>>()
        .unwrap();

    check_identity(&spec).await;

    // client: Check if client spec matches snapshot
    if config_modules.len() == 1 {
        let config = &config_modules[0];

        let client = print_schema(
            (Blueprint::try_from(config)
                .context(format!("file: {}", spec.path.to_str().unwrap()))
                .unwrap())
            .to_schema(),
        );

        let formatted = tailcall_prettier::format(client, &Parser::Gql)
            .await
            .unwrap();
        let snapshot_name = format!("{}_client", spec.safe_name);

        insta::assert_snapshot!(snapshot_name, formatted);
    }

    // run query tests
    run_query_tests_on_spec(spec, config_modules, mock_http_client).await;
}

pub async fn load_and_test_execution_spec(path: &Path) -> anyhow::Result<()> {
    let contents = fs::read_to_string(path)?;
    let spec: ExecutionSpec = ExecutionSpec::from_source(path, contents)
        .await
        .with_context(|| path.display().to_string())?;

    match spec.runner {
        Some(Annotation::Skip) => {
            println!("{} ... {}", spec.path.display(), "skipped".blue());
        }
        None => test_spec(spec).await,
    }

    Ok(())
}

async fn run_test(
    app_ctx: Arc<AppContext>,
    request: &APIRequest,
) -> anyhow::Result<http::Response<Body>> {
    let body = request
        .body
        .as_ref()
        .map(|body| Body::from(body.to_bytes()))
        .unwrap_or_default();

    let method = request.method.clone();
    let headers = request.headers.clone();
    let url = request.url.clone();
    let req = headers
        .into_iter()
        .fold(
            Request::builder()
                .method(method.to_hyper())
                .uri(url.as_str()),
            |acc, (key, value)| acc.header(key, value),
        )
        .body(body)?;

    // TODO: reuse logic from server.rs to select the correct handler
    if app_ctx.blueprint.server.enable_batch_requests {
        handle_request::<GraphQLBatchRequest>(req, app_ctx).await
    } else {
        handle_request::<GraphQLRequest>(req, app_ctx).await
    }
}
