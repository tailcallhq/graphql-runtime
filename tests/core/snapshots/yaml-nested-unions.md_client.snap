---
source: tests/core/spec.rs
expression: formatted
---
type Query {
  testVar0(u: T1Input!): U
  testVar1(u: T2Input!): U
  testVar2(u: T3Input!): U
  testVar3(u: T4Input!): U
  testVar4(u: T5Input!): U
}

type T1 {
  t1: String
}

input T1Input {
  t1: String
}

type T2 {
  t2: Int
}

input T2Input {
  t2: Int
}

type T3 {
  t3: Boolean
  t33: Float!
}

input T3Input {
  t3: Boolean
  t33: Float!
}

type T4 {
  t4: String
}

input T4Input {
  t4: String
}

type T5 {
  t5: Boolean
}

input T5Input {
  t5: Boolean
}

union U = T1 | T2 | T3 | T4 | T5

schema {
  query: Query
}
